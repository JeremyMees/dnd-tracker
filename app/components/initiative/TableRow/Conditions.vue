<script setup lang="ts">
import { useConditionsListing } from '~~/queries/open5e'

const props = defineProps<{
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}>()

type Condition = InitiativeSheetRow['conditions'][0]

const selected = ref<Condition[]>([])
const popoverOpen = ref<boolean>(false)

const { data: conditions, isPending } = await useConditionsListing()

watch(popoverOpen, open => selected.value = open ? props.item.conditions : [])

function removeCondition(name: string): void {
  if (!props.sheet) return

  const rows = props.sheet.rows

  props.update({
    rows: rows.map(row => row.id === props.item.id
      ? { ...row, conditions: row.conditions.filter(r => r.name !== name) }
      : row,
    ),
  })
}

function updateCondition(conditions: Condition[]): void {
  if (!props.sheet) return

  const index = getCurrentRowIndex(props.sheet, props.item.id)
  const rows = [...props.sheet.rows]

  if (index === -1 || !rows[index]) return

  rows[index] = { ...rows[index], conditions }

  props.update({ rows })

  popoverOpen.value = false
}

function toggleSelected(item: Open5eItem | Condition): void {
  const arr = [...selected.value]
  const index: number = arr.findIndex(s => s.name === item.name)

  const condition: Condition = {
    name: item.name,
    desc: typeof item.desc === 'string' ? item.desc : (item.desc as any)?.en || '',
    ...(item.level !== undefined && {
      level: typeof item.level === 'number' ? item.level : Number(item.level) || undefined,
    }),
    ...(('hasLevels' in item) && { hasLevels: item.hasLevels }),
  }

  if (index === -1) arr.push(condition)
  else arr.splice(index, 1)

  selected.value = arr
}

function listFromText(text: string, exhaustion: boolean = false): string[] {
  return exhaustion
    ? text
        .replace('*', '')
        .split(/\|\s\d+\s+\|/g)
        .slice(1)
        .map(bullet => bullet.split('|')[0])
        .filter((s): s is string => s !== undefined)
    : text.replace('*', '').split(/\s\*\s/g)
}
</script>

<template>
  <div
    v-if="item.type !== 'lair'"
    class="flex items-center gap-2"
  >
    <UiPopover v-model:open="popoverOpen">
      <UiPopoverTrigger as-child>
        <button
          :disabled="isPending"
          :class="{ 'cursor-progress': isPending }"
          class="h-[27px] flex flex-col justify-center"
        >
          <Icon
            name="tabler:plus"
            class="size-5 min-w-5 text-foreground/10"
            aria-hidden="true"
          />
        </button>
      </UiPopoverTrigger>
      <UiPopoverContent>
        <UiPopoverHeader>
          <UiPopoverTitle>
            {{ $t('pages.encounter.update.conditions') }}
          </UiPopoverTitle>
        </UiPopoverHeader>
        <div class="flex flex-wrap gap-2">
          <UiBadge
            v-for="condition in conditions"
            :key="condition.name"
            :variant="selected.map(s => s.name).includes(condition.name) ? 'default' : 'outline'"
            class="cursor-pointer"
            @click="toggleSelected(condition)"
          >
            {{ condition.name }}
          </UiBadge>
        </div>
        <div class="flex justify-end mt-4">
          <button
            class="btn-foreground"
            :aria-label="$t('actions.update')"
            @click="updateCondition(selected)"
          >
            {{ $t('actions.update') }}
          </button>
        </div>
      </UiPopoverContent>
    </UiPopover>
    <div
      v-if="item.conditions.length"
      class="flex flex-wrap justify-center md:justify-start gap-1"
    >
      <UiPopover
        v-for="condition in item.conditions"
        :key="condition.name"
      >
        <UiPopoverTrigger>
          <UiBadge class="whitespace-nowrap">
            {{ condition.name }} {{ condition.level ? `(${condition.level})` : '' }}
          </UiBadge>
        </UiPopoverTrigger>
        <UiPopoverContent>
          <UiPopoverHeader>
            <UiPopoverTitle>{{ condition.name }}</UiPopoverTitle>
          </UiPopoverHeader>
          <template v-if="condition.desc">
            <ul class="mx-6">
              <li
                v-for="bullet in listFromText(condition.desc, condition.name === 'Exhaustion')"
                :key="bullet"
                class="text-sm pb-1"
                :class="[condition.name === 'Exhaustion' ? 'list-decimal' : 'list-disc']"
              >
                {{ bullet }}
              </li>
            </ul>
          </template>
          <UiNumberField
            v-if="condition.hasLevels"
            id="level"
            :default-value="condition.level || 1"
            :min="1"
            :max="6"
            class="w-[100px] mt-4"
            @update:model-value="(level) => {
              const updatedConditions = [...item.conditions].map(c => c.name === condition.name ? { ...c, level } : c)
              updateCondition(updatedConditions)
            }"
          >
            <UiLabel for="level">
              {{ $t('general.level') }}
            </UiLabel>
            <UiNumberFieldContent>
              <UiNumberFieldDecrement />
              <UiNumberFieldInput />
              <UiNumberFieldIncrement />
            </UiNumberFieldContent>
          </UiNumberField>
          <div class="flex justify-end mt-4">
            <button
              :aria-label="$t('actions.remove')"
              class="btn-destructive"
              @click="removeCondition(condition.name)"
            >
              {{ $t('actions.remove') }}
            </button>
          </div>
        </UiPopoverContent>
      </UiPopover>
    </div>
  </div>
</template>
