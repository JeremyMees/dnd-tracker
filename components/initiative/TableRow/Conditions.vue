<script setup lang="ts">
defineEmits<{ openConditions: [] }>()

const props = defineProps<{
  item: InitiativeSheetRow
  sheet: InitiativeSheet
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable>) => Promise<void>
}>()

function removeEffect(name: string): void {
  const rows = props.sheet.rows

  props.update({
    rows: rows.map(row => row.id === props.item.id
      ? { ...row, conditions: row.conditions.filter(r => r.name !== name) }
      : row,
    ),
  })
}

function updateEffect(condition: InitiativeSheetRow['conditions'][0]): void {
  const index = getCurrentRowIndex(props.sheet, props.item.id)
  const rows = props.sheet.rows

  if (index === -1) return

  rows[index].conditions = rows[index].conditions.map(r => r.name === condition.name ? condition : r)

  props.update({ rows })
}
</script>

<template>
  <td>
    <div
      v-if="item.conditions.length"
      class="flex flex-wrap justify-center md:justify-start gap-1"
    >
      <Tag
        v-for="condition in item.conditions"
        :key="condition.name"
        removable
        :condition="condition"
        @update="updateEffect($event)"
        @remove="removeEffect($event)"
      />
    </div>
    <button
      v-else
      class="flex items-center gap-x-1"
      @click="$emit('openConditions')"
    >
      <Icon
        name="tabler:plus"
        class="size-5 min-w-5 text-secondary"
        aria-hidden="true"
      />
    </button>
  </td>
</template>
