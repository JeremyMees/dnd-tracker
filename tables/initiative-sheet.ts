import { createColumnHelper, type InitialTableState, type Row } from '@tanstack/vue-table'
import { actionsTable, expandButton, iconButton } from './generate-functions'

import {
  InitiativeTableRowName,
  InitiativeTableRowHealth,
  InitiativeTableRowActions,
  InitiativeTableRowConditions,
  InitiativeTableRowNotes,
  InitiativeTableRowInit,
  InitiativeTableRowAc,
  InitiativeTableRowDeathSaves,
  InitiativeTableRowConcentration,
  InitiativeTableRowModify,
} from '#components'

const columnHelper = createColumnHelper<InitiativeSheetRow>()

interface ColumnOptions {
  sheet: ComputedRef<InitiativeSheet | undefined>
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable>) => Promise<void>
  openQuickInitModal: () => void
}

export function generateColumns({ sheet, update, openQuickInitModal }: ColumnOptions) {
  const { t } = useI18n()

  return [
    columnHelper.display({
      enableGlobalFilter: false,
      id: 'index',
      cell: ({ row }) => h('div', { class: 'flex items-center gap-x-2 max-w-20' }, [
        h('span', { class: 'text-muted-foreground ml-1' }, row.index + 1),
        expandButton({
          content: t(`actions.${row.getIsExpanded() ? 'hide' : 'show'}`),
          expanded: row.getIsExpanded(),
          cb: () => row.toggleExpanded(),
        }),
      ]),
    }),
    columnHelper.accessor('name', {
      header: t('components.encounterTable.headers.name'),
      cell: ({ row }) => h(InitiativeTableRowName, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('initiative', {
      header: () => h('button', { class: 'flex items-center gap-x-2' }, [
        t('components.encounterTable.headers.init'),
        iconButton({
          icon: 'tabler:bolt-filled',
          content: t('components.encounterTable.quick'),
          color: 'warning',
          cb: () => openQuickInitModal(),
          disabled: !sheet.value?.rows.length,
          iconColor: 'text-warning',
        }),
      ]),
      cell: ({ row }) => h(InitiativeTableRowInit, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('health', {
      header: t('components.encounterTable.headers.health'),
      cell: ({ row }) => h(InitiativeTableRowHealth, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('ac', {
      header: t('components.encounterTable.headers.ac'),
      cell: ({ row }) => h(InitiativeTableRowAc, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      id: 'manage',
      header: t('components.encounterTable.headers.manage'),
      cell: ({ row }) => h(InitiativeTableRowActions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('conditions', {
      header: t('components.encounterTable.headers.conditions'),
      cell: ({ row }) => h(InitiativeTableRowConditions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('note', {
      header: t('components.encounterTable.headers.note'),
      cell: ({ row }) => h(InitiativeTableRowNotes, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('deathSaves', {
      header: t('components.encounterTable.headers.deathSaves'),
      cell: ({ row }) => h(InitiativeTableRowDeathSaves, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('concentration', {
      header: t('components.encounterTable.headers.concentration'),
      cell: ({ row }) => h(InitiativeTableRowConcentration, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      enableGlobalFilter: false,
      enableSorting: false,
      id: 'actions',
      cell: ({ row }) => h(InitiativeTableRowModify, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
  ]
}

export function expandedMarkup(row: Row<InitiativeSheetRow>) {
  return actionsTable(row.original)
}

export const initialState: InitialTableState = {}
