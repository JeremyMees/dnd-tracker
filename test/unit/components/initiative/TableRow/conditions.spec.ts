import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi, beforeEach } from 'vitest'
import Conditions from '~/components/initiative/TableRow/Conditions.vue'
import { sheet } from '~~/test/unit/fixtures/initiative-sheet'
import conditions from '~~/test/unit/fixtures/conditions.json'

interface Props {
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}

type Condition = InitiativeSheetRow['conditions'][0]

const mockUpdate = vi.fn()

const props: Props = {
  item: sheet.rows[0]!,
  sheet,
  update: mockUpdate,
}

// Mock the useConditionsListing composable
vi.mock('~~/queries/open5e', () => ({
  useConditionsListing: () => ({
    data: ref(conditions),
    isPending: ref(false),
  }),
}))

describe('Initiative table row conditions', async () => {
  beforeEach(() => mockUpdate.mockClear())

  it('Should match snapshot', async () => {
    const component = await mountSuspended(Conditions, { props })
    expect(component.html()).toMatchSnapshot()
  })

  it('Should not render for lair type items', async () => {
    const component = await mountSuspended(Conditions, {
      props: {
        ...props,
        item: { ...props.item, type: 'lair' },
      },
    })

    expect(component.find('div').exists()).toBeFalsy()
  })

  it('Should always show add condition button', async () => {
    const component = await mountSuspended(Conditions, { props })

    expect(component.find('[data-test-trigger]').exists()).toBeTruthy()
  })

  it('Should show selected conditions', async () => {
    const component = await mountSuspended(Conditions, {
      props: {
        ...props,
        item: {
          ...props.item,
          conditions: [
            conditions[0] as Condition,
            conditions[1] as Condition,
          ],
        },
      },
    })

    expect(component.find('[data-test-conditions]').exists()).toBeTruthy()

    const badges = component.findAll('[data-test-badge]')

    expect(badges.length).toBe(2)
    expect(badges[0]!.text()).toBe(conditions[0]!.name)
    expect(badges[1]!.text()).toBe(conditions[1]!.name)
  })

  it('Should not show conditions when no conditions are selected', async () => {
    const component = await mountSuspended(Conditions, {
      props: {
        ...props,
        item: {
          ...props.item,
          conditions: [],
        },
      },
    })

    expect(component.find('[data-test-conditions]').exists()).toBeFalsy()
  })

  it('Should display the level of the condition if available', async () => {
    const component = await mountSuspended(Conditions, {
      props: {
        ...props,
        item: {
          ...props.item,
          conditions: [{
            ...conditions[0] as Condition,
            level: 2,
          }],
        },
      },
    })

    expect(component.find('[data-test-badge]').text()).toBe(`${conditions[0]!.name} (2)`)
  })
})
