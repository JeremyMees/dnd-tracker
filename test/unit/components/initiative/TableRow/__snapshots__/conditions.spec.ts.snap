// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Initiative table row conditions > Should match snapshot 1`] = `
"<div class="flex items-center gap-2"><button data-test-trigger="" class="h-[27px] flex flex-col justify-center" id="reka-popover-trigger-v-0-0-0" type="button" aria-haspopup="dialog" aria-expanded="false" aria-controls="" data-state="closed"><span class="iconify i-tabler:plus size-5 min-w-5 text-foreground/10" aria-hidden="true"></span></button>
  <!--teleport start-->
  <!--teleport end-->
  <div data-test-conditions="" class="flex flex-wrap justify-center md:justify-start gap-1"><button id="reka-popover-trigger-v-0-0-1" type="button" aria-haspopup="dialog" aria-expanded="false" aria-controls="" data-state="closed">
      <div class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-foreground bg-foreground/50 text-background hover:bg-foreground/60 whitespace-nowrap" data-test-badge="">Paralyzed </div>
    </button>
    <!--teleport start-->
    <!--teleport end-->
  </div>
</div>"
`;
